site/content/3.12/aql/execution-and-performance/caching-query-plans.md

```diff
@@ -0,0 +1,136 @@
+```
+
+See the [HTTP API](../../develop/http-api/queries/aql-query-plan-cache.md) for details. <!-- TODO: link to cursor API or query cache API? -->
+{{< /tab >}}
+
+{{< /tabs >}}
+
+## Cache eligibility
+
+<!-- TODO
+LIMIT offset/count values must be constant numeric values - only raised if plan caching is on
+-->
+If plan caching is enabled for query, the eligibility for being cached is checked
+first. Any queries that have any of the following query options set are not
+eligible for plan caching:
+
+- `allPlans`
+- `optimizer.rules`
+- `explainRegisters` (internal)
+- `inaccessibleCollections` (internal)
+- `shardIds` (internal)
+
+Additionally, a query is not eligible for plan caching if it uses attribute 
+name bind parameters (e.g. `FILTER doc.@attributeName == ...`) or when using
+value bind parameters in any of the following places:
+- Specifying the depths for traversals or path queries
+  (e.g. `FOR v, e IN @min..@max OUTBOUND ...`)
+- Referring to a named graph (e.g. `GRAPH @graphName`)
+- Referring to edge collections used in traversals or path queries 
+  (e.g. `FOR v, e IN 1..2 OUTBOUND 'v/0' @@edgeColl ...`)
+
+If a query produces any warnings during parsing or query plan optimization,
+it is also not eligible for plan caching.
+
+Query plans are also not eligible for caching if they contain one of the
+following execution node types in the plan:
+- `SingleRemoteOperationNode` (cluster only)
+- `MultipleRemoteModificationNode` (cluster only)
+- `UpsertNode`, i.e. the AQL `UPSERT` operation
+
+If a query is eligible for plan caching, the plan cache is checked using
+the exact same query string and set of collection bind parameter values.
+A cached plan entry is only considered identical to the current query if the
+query strings are identical byte for byte and the set of collection bind 
+parameters is exactly the same (bind parameter names and bind parameter 
+values).
+
+If no plan entry can be found in the plan cache, the query is planned and 
+optimized as usual, and the cached plan is inserted into the plan cache.
+Repeated executions of the same query (same query string and using the same 
+set of collection bind parameters) then makes use of the cached plan
+entry, potentially with different value bind parameters.
+
+The following query options are ignored when a cached plan is used:
+- `joinStrategyType`
+- `maxDNFConditionMembers`
+- `maxNodesPerCallstack`
+
+Whenever a query uses a plan from the plan cache, the query
+result includes a `planCacheKey` attribute on the top level when
+executing, explaining, or profiling a query. The explain and profiling
+outputs also indicate when a cached query plan was used, showing
+`plan cache key: ...` at the top.
+
+## Cache invalidation
+
+The query plan cache is organized per database. It gets invalidated at the
+following events:
+- An existing collection gets dropped or renamed, or the properties of an
+  existing collection are modified
```

Comment:
But only if the affected collection is in the list of data sources for the cached query plan?

Reply:
Note to self: Check again whether the entire cache is now cleared on any of these events, including e.g. changing the schema, computed values, or cacheEnabled of a collection in single server.

Reply:
In single server, the cache isn't necessarily cleared entirely but it depends on the operation. I suggest we add a vague remark to the docs that operations may only remove affected entries in single server.